core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala

```diff
@@ -0,0 +1,89 @@
+package org.broadinstitute.dsde.rawls.snapshot
+
+import bio.terra.datarepo.model.{SnapshotModel, CloudPlatform => SnapshotCloudPlatform}
+import org.broadinstitute.dsde.rawls.RawlsException
+import org.broadinstitute.dsde.rawls.model.WorkspaceCloudPlatform.WorkspaceCloudPlatform
+import org.broadinstitute.dsde.rawls.model.{Workspace, WorkspaceCloudPlatform}
+
+import scala.jdk.CollectionConverters.CollectionHasAsScala
+
+object SnapshotReferenceCreationValidator {
+  def constructor(workspaceContext: Workspace, snapshot: SnapshotModel): SnapshotReferenceCreationValidator =
+    new SnapshotReferenceCreationValidator(workspaceContext, snapshot)
+}
+
+// thrown to disallow creation of a snapshot reference on an unsupported platform
+class UnsupportedPlatformException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across cloud boundaries
+class PlatformBoundaryException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across a protected data boundary
+class ProtectedDataException(message: String) extends RawlsException(message)
+
+class SnapshotReferenceCreationValidator(val workspaceContext: Workspace, val snapshot: SnapshotModel) {
+
+  // TODO: get this information from a more authoritative source rather than relying on the hardcoded bucket prefix
+  private def isWorkspaceProtected: Boolean = workspaceContext.bucketName.startsWith("fc-secure")
+
+  private def isSnapshotProtected: Boolean = snapshot.getSource.asScala.exists(_.getDataset.isSecureMonitoringEnabled)
+
+  private def snapshotPlatform: SnapshotCloudPlatform = {
+    val snapshotSourceDatasets = snapshot.getSource.asScala
+    val datasetCloudPlatforms = snapshotSourceDatasets.map(_.getDataset.getCloudPlatform).filter(_ != null).toSet
+
+    // assume GCP if datasets don't come from a source explicitly marked with another platform
+    if (datasetCloudPlatforms.isEmpty) {
+      return SnapshotCloudPlatform.GCP
+    }
+
+    // fail if datasets sources somehow span multiple clouds
+    if (datasetCloudPlatforms.size > 1) {
+      throw new IllegalStateException(
+        s"snapshot ${snapshot.getId} unexpectedly has more than one cloudPlatform: $datasetCloudPlatforms"
+      )
+    }
+
+    // fail if dataset sourced from something other than GCP aren't all explicitly marked as such
+    // as this might indicate co-mingled GCP (blank) & non-GCP datasets
+    val onlyCloudPlatform = datasetCloudPlatforms.take(1).last
+    onlyCloudPlatform match {
+      case SnapshotCloudPlatform.GCP => onlyCloudPlatform
+      case _ =>
+        if (snapshotSourceDatasets.forall(_.getDataset.getCloudPlatform == onlyCloudPlatform)) onlyCloudPlatform
+        else
+          throw new IllegalStateException(
+            s"snapshot ${snapshot.getId} unexpectedly has sources from more than one cloudPlatform"
+          )
```

Comment:
how would we ever reach this second case? We've already run `toSet` which eliminates duplicates, and we already have a check for `datasetCloudPlatforms.size > 1` … this feels redundant.

If you combine a bunch of these comments, I think you can simplify this method to:
```scala
  private def snapshotPlatform: SnapshotCloudPlatform = {
    val snapshotSourceDatasets = snapshot.getSource.asScala
    val datasetCloudPlatforms = snapshotSourceDatasets.flatMap(x => Option(x.getDataset.getCloudPlatform)).toSet

    // fail if datasets sources somehow span multiple clouds
    if (datasetCloudPlatforms.size > 1) {
      throw new IllegalStateException(
        s"snapshot ${snapshot.getId} unexpectedly has more than one cloudPlatform: $datasetCloudPlatforms"
      )
    }

    // we're left with either zero or one cloud platform from the snapshot sources
    datasetCloudPlatforms.headOption match {
      // assume GCP if datasets don't come from a source explicitly marked with another platform
      case None => SnapshotCloudPlatform.GCP
      case Some(platform) => platform
    }
  }
```

Reply:
You're right this is really gross and where I thought I might be over-engineering... thanks for pulling me back 😆 

By way of explanation, I think this got all bloated because I wasn't sure if I could assume `getCloudPlatform` was always set for every dataset, and wanted to handle the case where there were blanks (defaulted to GCP) co-mingled with non-blanks (which were Azure), and fail in that situation.  Hopefully @nmalfroy can confirm that it's always present and then we can eliminate this!

Reply:
See newly simplified `WrappedSnapshot` which de-uglies things a lot.